File: Minimum_Index_Sum_of_Two_Lists.go
package main

import (
	"fmt"
)

func main() {

	// list1 := []string{"Shogun", "Piatti", "Tapioca Express", "Burger King", "KFC"}
	// list2 := []string{"Piatti", "The Grill at Torrey Pines", "Hungry Hunter Steakhouse", "Shogun"}

	list1 := []string{"happy", "sad", "good"}
	list2 := []string{"sad", "happy", "good"}

	// list1 := []string{"Shogun", "Tapioca Express", "Burger King", "KFC"}
	// list2 := []string{"Piatti", "The Grill at Torrey Pines", "Hungry Hunter Steakhouse", "Shogun"}

	fmt.Println(findRestaurant(list1, list2))
}
func findRestaurant(list1, list2 []string) []string {
	l1 := make(map[string]int)

	for i, s := range list1 {
		l1[s] = i
	}
	out := 3000
	var res []string
	for i, s := range list2 {
		if _, ok := l1[s]; ok {
			sum := i + l1[s]
			if sum < out {
				out = sum
				res = []string{s}
			} else if out == sum {
				res = append(res, s)
			}

		}
	}

	return res
}

File: Minimum_Index_Sum_of_Two_Lists.rs
use std::collections::HashMap;

fn main(){
    let list1:Vec<String> = ["Shogun".to_string(),"Tapioca Express".to_string(),"Burger King".to_string(),"KFC".to_string()].to_vec();
    let list2:Vec<String> = ["Piatti".to_string(),"The Grill at Torrey Pines".to_string(),"Hungry Hunter Steakhouse".to_string(),"Shogun".to_string()].to_vec();
    println!("{:?} ",find_restaurant(list1,list2)); 
}


fn find_restaurant(list1:Vec<String>,list2:Vec<String>) -> Vec<String>{
    let mut l1= HashMap::new();    
    for (i,s) in list1.iter().enumerate(){
        l1.insert(s,i);
    }
    
    let mut v1:Vec<String> = Vec::new();
    let mut min = list1.len() + list2.len();
    for (i,s) in list2.iter().enumerate(){
        if let Some(idx) = l1.get(s){
            if i+idx == min{
                v1.push(s.clone());
            }else if i + idx < min {
            v1.clear();
            v1.push(s.clone());
            min = i+idx;
            }
        }
    }


    return v1;
}